include/igl/viewer/Viewer.cpp

```diff
@@ -355,39 +468,148 @@ namespace viewer
     }
   }
 
+  IGL_INLINE unsigned int Viewer::add_mesh(const std::string& id)
+  {
+    data_buffer.push_back(ViewerData());
+    
+    data_ids.push_back(id);
+    
+    opengl.push_back(OpenGL_state());
+
+    if(window != nullptr)
+    {
+      opengl[opengl.size()-1].init();
+#if defined(IGL_VIEWER_WITH_NANOGUI) && defined(IGL_VIEWER_WITH_NANOGUI_MULTIMESH)
+      currentDataCB->setItems(data_ids);
+      screen->performLayout();
+#endif
+    }
+    else
+    {
+      set_active_mesh(opengl.size()-1);
+    }
+
+    return opengl.size()-1;
+  }
+
+  IGL_INLINE unsigned int Viewer::add_mesh(const char* mesh_file_name,const std::string& id)
+  {
+    unsigned int mid = add_mesh(id);
+    load_mesh_from_file(mesh_file_name);
+
+#if defined(IGL_VIEWER_WITH_NANOGUI) && defined(IGL_VIEWER_WITH_NANOGUI_MULTIMESH)
+    if(window != nullptr)
+    {
+      screen->performLayout();
+    }
+#endif
+
+    return mid;
+  }
+
+  IGL_INLINE ViewerData& Viewer::get_mesh(unsigned int data_id)
+  {
+    assert(data_buffer.size() > data_id && "data_id out of range");
+
+    if(data_id == active_data_id)
+      return this->data;
+    else
+      return data_buffer[data_id];
+  }
+
+  IGL_INLINE bool Viewer::remove_mesh(unsigned int data_id)
+  {
+    assert(data_buffer.size() > data_id && "data_id out of range");
+    assert(active_data_id != data_id && "active mesh cannot be removed!");
+    
+    if(active_data_id > data_id)
+      active_data_id--;
+
+    data_buffer.erase(data_buffer.begin()+data_id);
+    data_ids.erase(data_ids.begin()+data_id);
+    opengl[data_id].free();
+    opengl.erase(opengl.begin()+data_id);
+
+#if defined(IGL_VIEWER_WITH_NANOGUI) && defined(IGL_VIEWER_WITH_NANOGUI_MULTIMESH)
+    if(window != nullptr)
+    {
+      currentDataCB->setItems(data_ids);
+      screen->performLayout();
+    }
+#endif
+
+    return true;
+  }
+
+  IGL_INLINE unsigned int Viewer::get_active_mesh()
+  {
+    return active_data_id;
+  }
+
+  IGL_INLINE bool Viewer::set_active_mesh(unsigned int data_id)
+  {
+    assert(data_buffer.size() > data_id && "data_id out of range");
+
+    data_buffer[active_data_id] = ViewerData(data);
+    active_data_id = data_id;
+    data = data_buffer[active_data_id];
```

Comment:
That's correct, that's the only way to not have to change the current "interface" of the viewer where a lot of code is using the member data directly. Or do you have another suggestion?